### DISCUSSION TITLE
How to run the Raspberry pi 4B hardware on WSL

### DISCUSSION THREAD

**Discussion by Nainikas** [nasa/fprime#2761] [General]
![Screenshot 2024-06-06 234244](https://github.com/nasa/fprime/assets/55590634/9aa2e0fa-51bb-4da2-9cd1-c502ffb4304e)
I’m on Windows 11 and I have installed WSL 2. The tutorials which do not need hardware seem to work fine but the LED Blinker kept giving me port errors despite the codes being ok. Maybe it’s because the WSL and ARM requirements do not match? Any idea on how to resolve it?

**Comment by kevin-f-ortega** [comment]
What error are you seeing? I suspect you need to enable GPIO on the pi.

Run this on the pi 

sudo raspi-config
Select Interface options
Select Remote GPIO
Select Finished
Reboot the pi

Then try running the FSW again

**Reply by Nainikas** [reply]
![Screenshot 2024-06-06 234244](https://github.com/nasa/fprime/assets/55590634/7dfb96a3-a385-473a-a7a0-b06b18c641fb)

**Reply by Nainikas** [reply]
I did enable the GPIO, yet the error occurs.

**Reply by LeStarch** [reply]
This error can occur in 3 cases:
1. The GPIO ports are not enabled. (You checked this)
2. The user does not have permission to `gpio`.  Make sure your user is in the `gpio` group or run as `root`
3. Another process is using that GPIO.

**Reply by Nainikas** [reply]
I tried those methods but still failed. I'm not sure where I'm going wrong. This is what I end up getting.
![image](https://github.com/nasa/fprime/assets/55590634/1f3574dd-b72b-4eb4-8e96-14ab5e8653a6)

**Reply by Nainikas** [reply]
Update: I ran a python script on the Raspberry Pi with GPIO 13 as output pin and it works fine. I am guessing it means that my GPIO permissions and settings are OK. Please help me resolve the issue with FPrime.

**Reply by thomas-bc** [reply]
Could you link us to the code you are running? Also can you clarify whether this error is happening on the RaspberryPi or on WSL2 ?

Here is a reference you can use: https://github.com/fprime-community/fprime-workshop-led-blinker
This works on my RaspberryPi on my end

**Reply by Nainikas** [reply]
I’m following the tutorial, https://fprime-community.github.io/fprime-workshop-led-blinker/ and I’m on WSL2. Although, the LedBlinkerTopology file is a bit different than what got generated for me.

**Reply by Nainikas** [reply]
Update: I tried the tutorial again on Linux OS but the same error occurs.

**Reply by thomas-bc** [reply]
Could you link us to the code you're running (published to GitHub)? Without this, it's really hard to say what's wrong
Can you also try to use the reference repo I linked and see if that succeeds or not? See below for sample commands

```
pip install fprime-bootstrap
fprime-bootstrap clone https://github.com/fprime-community/fprime-workshop-led-blinker
<activate venv>
cd fprime-workshop-led-blinker/LedBlinker
fprime-util generate
fprime-util build -j16
```

**Reply by Nainikas** [reply]
https://github.com/Nainikas/Led-Blinker-Tutorial/tree/master

**Reply by thomas-bc** [reply]
This looks good to me...

Since you're likely not going to actually connect anything to the GPIOs, you can do 
```
fprime-util generate -DFPRIME_USE_STUBBED_DRIVERS=ON
```
This will cause the GPIO operations to be no-ops, and it should work for testing things out.

Do you not have a RaspberryPi ? The tutorial was really designed to be ran on a RPI, so I wouldn't worry too much about this GPIO error if you don't have one.

**Reply by Nainikas** [reply]
Thank you for checking my work.

I do have a Raspberry Pi 4b which I was hoping to try it on. Any recommendations for that?

**Reply by thomas-bc** [reply]
Ah yes! You should follow the instructions here https://fprime-community.github.io/fprime-workshop-led-blinker/docs/running-on-hardware.html
It should guide you through how to get cross compilers and such

**Reply by Nainikas** [reply]
I get those errors when I do the cross-compilation

**Reply by thomas-bc** [reply]
I don't quite understand your workflow. Are you cross compiling and then uploading to the raspberrypi ? (that is, following this tutorial https://github.com/nasa/fprime/blob/devel/docs/Tutorials/CrossCompilationSetup/ArmLinuxTutorial.md)

**Reply by Nainikas** [reply]
Yes, I am. Is that not the right way?

**Reply by Nainikas** [reply]
Should I do it any other way?

**Reply by thomas-bc** [reply]
No this is the right way. At this point, I don't know what's wrong.

Maybe clone the project on the RaspberryPi and compile it natively there?

**Reply by Nainikas** [reply]
I did that too. It gives the same error! :(

**Reply by Nainikas** [reply]
Thank you for all your help! I got a response and solution from here, https://github.com/fprime-community/fprime-workshop-led-blinker/issues/67

Looks like I had to change the GPIO addressing.
